src/scheduler/server_info.cpp

```diff
@@ -2107,156 +2072,131 @@ check_resv_running_on_node(resource_resv *resv, const void *arg)
 	return 0;
 }
 
-/**
- * @brief
- * 		dup_server_info - duplicate a server_info struct
- *
- * @param[in]	osinfo	-	the struct to copy
- *
- * @return	duplicated server_info
- * @retval	NULL	: something wrong!
- *
- * @par MT-Safe:	no
- */
-server_info *
-dup_server_info(server_info *osinfo)
-{
-	server_info *nsinfo;		/* scheduler internal form of server info */
-	unsigned int i;
-
-	if (osinfo == NULL)
-		return NULL;
-
-	/* duplicate the server information */
-	if ((nsinfo = new_server_info(0)) == NULL)
-		return NULL;
-
-	if (osinfo->fstree != NULL) {
-		nsinfo->fstree = new fairshare_head(*osinfo->fstree);
-		if (nsinfo->fstree == NULL) {
-			free_server(nsinfo);
-			return NULL;
-		}
-	}
-	nsinfo->has_mult_express = osinfo->has_mult_express;
-	nsinfo->has_soft_limit = osinfo->has_soft_limit;
-	nsinfo->has_hard_limit = osinfo->has_hard_limit;
-	nsinfo->has_user_limit = osinfo->has_user_limit;
-	nsinfo->has_all_limit = osinfo->has_all_limit;
-	nsinfo->has_grp_limit = osinfo->has_grp_limit;
-	nsinfo->has_proj_limit = osinfo->has_proj_limit;
-	nsinfo->has_multi_vnode = osinfo->has_multi_vnode;
-	nsinfo->has_prime_queue = osinfo->has_prime_queue;
-	nsinfo->has_nonprime_queue = osinfo->has_nonprime_queue;
-	nsinfo->has_ded_queue = osinfo->has_ded_queue;
-	nsinfo->has_nodes_assoc_queue = osinfo->has_nodes_assoc_queue;
-	nsinfo->node_group_enable = osinfo->node_group_enable;
-	nsinfo->eligible_time_enable = osinfo->eligible_time_enable;
-	nsinfo->provision_enable = osinfo->provision_enable;
-	nsinfo->power_provisioning = osinfo->power_provisioning;
-	nsinfo->use_hard_duration = osinfo->use_hard_duration;
-	nsinfo->pset_metadata_stale = osinfo->pset_metadata_stale;
-	nsinfo->name = osinfo->name;
-	nsinfo->liminfo = lim_dup_liminfo(osinfo->liminfo);
-	nsinfo->server_time = osinfo->server_time;
-	nsinfo->res = dup_resource_list(osinfo->res);
-	nsinfo->alljobcounts = dup_counts_map(osinfo->alljobcounts);
-	nsinfo->group_counts = dup_counts_map(osinfo->group_counts);
-	nsinfo->project_counts = dup_counts_map(osinfo->project_counts);
-	nsinfo->user_counts = dup_counts_map(osinfo->user_counts);
-	nsinfo->total_alljobcounts = dup_counts_map(osinfo->total_alljobcounts);
-	nsinfo->total_group_counts = dup_counts_map(osinfo->total_group_counts);
-	nsinfo->total_project_counts = dup_counts_map(osinfo->total_project_counts);
-	nsinfo->total_user_counts = dup_counts_map(osinfo->total_user_counts);
-	nsinfo->node_group_key = osinfo->node_group_key;
-	nsinfo->nodesigs = dup_string_arr(osinfo->nodesigs);
-
-	nsinfo->policy = dup_status(osinfo->policy);
-
-	nsinfo->num_nodes = osinfo->num_nodes;
+// Copy constructor
+server_info::server_info(const server_info &osinfo)
+{
+	init_server_info();
```

Comment:
I wouldn't init everything just to copy over everything again.  Just keep in mind that the constructor hasn't been called, so you do actually need to set every member.